### ISSUE TITLE
Improve validation of planetary reference frames

### ISSUE THREAD

**Issue by astrojuanlu** [poliastro/poliastro#109]
In the early days of poliastro, a question was raised about the reference frames of bodies other than the Earth:

http://physics.stackexchange.com/q/69792/7641

The answer is that these are defined by the IAU Working Group on Cartographic Coordinates and Rotational Elements (WGCCRE).

http://astrogeology.usgs.gov/groups/IAU-WGCCRE

These frames should be taken into account when defining orbits around the bodies.

**Comment by Juanlu001** [comment]
The notion of a reference frame within this package is still very unclear and needs careful thought. Postponing this.

**Comment by Juanlu001** [comment]
In fact, right now the plotting code is the only one that makes use of reference frames, and it's done in a very artificial way.

https://github.com/poliastro/poliastro/blob/93d58de/src/poliastro/plotting.py#L63-L84

**Comment by Juanlu001** [comment]
See this for a starting point:

https://github.com/astropy/astropy/blob/v1.3/astropy/coordinates/builtin_frames/icrs_cirs_transforms.py#L287-L297

Perhaps this goes to Astropy itself? Note to self: open an issue.

**Comment by Juanlu001** [comment]
Quoting from the official report:

> Approximate expressions for these rotational elements with respect to the International Celestial Reference Frame (ICRF) (Ma et al. 1998) have been derived. The ICRF is the reference frame of the  International Celestial Reference System and is itself epochless.

https://astropedia.astrogeology.usgs.gov/download/Docs/WGCCRE/WGCCRE2009reprint.pdf

With this data, in theory it would be possible to implement inertial, planet-centered reference frames.

**Comment by Juanlu001** [comment]
We will probably implement this as part of Astropy, see https://github.com/astropy/astropy/issues/5898.

**Comment by Juanlu001** [comment]
Relevant: https://github.com/astropy/astropy/issues/6472

**Comment by Juanlu001** [comment]
@anhiga started working on this in #215.

**Comment by Juanlu001** [comment]
These should be validated against SpiceyPy, since SPICE has planetary reference frames defined:

https://naif.jpl.nasa.gov/pub/naif/toolkit_docs/C/req/frames.html#Appendix.%20%60%60Built%20in%27%27%20PCK-Based%20IAU%20Body-Fixed%20Reference%20Frames

**Comment by Juanlu001** [comment]
I'm blocking this until I have proper conda-forge recipes for SpiceyPy. And unfortunately that involves making them work on Windows...

https://github.com/conda-forge/staged-recipes/pull/4115

**Comment by Juanlu001** [comment]
Also, the Astropy effort is longer term, so I'm removing the `upstream` label.

**Comment by Juanlu001** [comment]
This would also help improving the overall coverage significantly: https://codecov.io/gh/poliastro/poliastro/tree/master/src/poliastro

**Comment by Juanlu001** [comment]
There are already conda-forge packages for SpiceyPy, but they don't work on Windows.

**Comment by Juanlu001** [comment]
Also, let's implement this here first, and then we can push this upstream and collaborate with the SunPy folks.

**Comment by Juanlu001** [comment]
I added general information about validation here https://github.com/poliastro/poliastro/wiki/Validation

**Comment by Juanlu001** [comment]
Started working on this: https://github.com/conda-forge/cspice-feedstock/pull/1

**Comment by Juanlu001** [comment]
Made some progress with cspice and spiceypy, but no Windows packages yet.

**Comment by Juanlu001** [comment]
cspice conda packages for Windows are there! :tada: I'm now unblocking this issue

**Comment by Juanlu001** [comment]
Postponing for the next release, this should be our focus.

**Comment by Juanlu001** [comment]
I learned that GMAT has Python scripting capabilities, so I'm updating the wiki to add that information. There are some caveats though: http://forums.gmatcentral.org/viewtopic.php?f=48&t=49258

**Comment by Juanlu001** [comment]
OK, I tried to get GMAT to work but I failed to understand some internal discrepancies with ephemeris data:

http://forums.gmatcentral.org/viewtopic.php?f=2&t=49271&p=52056

On the other hand, I didn't find a way to include the position of bodies different from Earth and Moon in STK:

http://help.agi.com/stk/index.htm#../Subsystems/dataProviders/Content/Scenario~Points(ICRF).htm

So I decided that I will go for SPICE and SpiceyPy.

**Comment by Juanlu001** [comment]
It turns out that SPICE frames only account for rotation and not translation. From the responses of [this Twitter thread](https://twitter.com/poliastro_py/status/1036316671072522244):

> I believe SPICE frames are only for orientation. Centering it at a body does not mean that transformations will include the translation. -- @jacobwilliams

> Jacob is correct. SPICE frames are just for rotation. You will have to evaluate the ephemeris with „spkezr“ and do the vector math yourself. -- @helgee

So, for the case of ICRS frames, the validation is trivial. However, it's worth it to use it for IAU body-fixed frames, but that will go in the following release.

**Comment by Juanlu001** [comment]
I discussed with @marcsit about this issue and he says that it's indeed possible to define a reference frame in SPICE with translations, not only orientation. I still haven't tested whether the PCK frames do, in fact, store the offset from the Solar System Barycenter. We need something like body-fixed frames, but parallel to the ICRS. What do you think?

**Comment by Juanlu001** [comment]
Removing my assignation from this issue, as someone might want to tackle it in GSOC.

**Comment by GorgiAstro** [comment]
Orekit supports frame translation, rotation or both, and has inertial and body-fixed frames for all planets + the Moon. Using the Python wrapper would enable to integrate Orekit in the test cases without having to rely on hard-coded data from GMAT/STK/etc. If it is of interest, I can start working on it.

This could be used for #451 as well.

**Comment by Juanlu001** [comment]
Hi @GorgiAstro, thanks for chiming in! In fact I didn't consider Orekit but that was an oversight on my part - it would be great to validate poliastro against it. The great @petrushy maintains the Orekit Python bindings https://gitlab.orekit.org/orekit-labs/python-wrapper, so this is definitely feasible! Could you please add Orekit here https://github.com/poliastro/poliastro/wiki/Validation?

There is an ongoing pull request to address #451 with SPICE, see #602. However, it would be great to have it validated against Orekit as well.

**Comment by GorgiAstro** [comment]
> Could you please add Orekit here https://github.com/poliastro/poliastro/wiki/Validation?
Done.

> There is an ongoing pull request to address #451 with SPICE, see #602. However, it would be great to have it validated against Orekit as well.
Okay, I will get started on test cases using Orekit-Python then. Besides, it is a good way for me to get used to working with poliastro. I created a separate ticket #644

**Comment by Juanlu001** [comment]
After a long silence, I plan to get back to the work @amsuhane and @GorgiAstro did during the (Northern) summer for poliastro 0.14. Before that, I would like to make a big refactor of the `bodies` module to remove the rotational elements from there and simplify the class structure, and possibly tackle #703 along the way.

**Comment by astrojuanlu** [comment]
This is finally being addressed in https://github.com/poliastro/validation/pull/24! poliastro is now validated 🚀
